### PR TITLE
Implement BIP47 Payment Code and notification address

### PR DESCRIPTION
**Pull request by kannapoix**
# Improvements made by this PR
Implement v1 of [BIP47](https://github.com/bitcoin/bips/blob/master/bip-0047.mediawiki#Abstract) Payment Code to bitcoinrb.  
This PR achieved following features.
- Generate Payment Code
- Decode Payment Code
- Generate notification address

# Required further works
- Notification transaction
- Sending transaction
- Whole v2 implementation

### CODE REVIEW

File: lib/bitcoin/payment_code.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    def notification_address
+      ext_pubkey.derive(0).addr
+    end
+
+    def self.from_payment_code_string(paymen_code_string)
+      hex = Bitcoin::Base58.decode(paymen_code_string)
+      row_payment_code = hex[0...-8]
+      raise ArgumentError, 'invalid checksum' unless Bitcoin.calc_checksum(row_payment_code) == hex[-8..-1]
+
```
Comment (azuchi) [inline lib/bitcoin/payment_code.rb:40]: The following checks are required in addition to the checksum.

* version is correct?(In this implementation, support only 0x01)
* whether x value is a point on secp256k1?
* sign value is supported?(0x02 or 0x03)?

File: lib/bitcoin/payment_code.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    attr_accessor :x_value
+
+    VERSION_BYTE = '47'
+    SUPPORT_VERSIONS = ['01']
+    SUPPORT_SIGNS = ['02', '03']
+
+    def initialize
+      @version = '01'
+      @features_bits = '00'
+      @sign = '02'
```
Comment (azuchi) [inline lib/bitcoin/payment_code.rb:14]: In this implementation, y coordinate of the compressed public key is always positive.

**Comment by kannapoix** [comment]
@azuchi 
Thank you for reviewing. I fixed the points you suggested and also made additional fix.

**Comment by azuchi** [comment]
Thanks @kannapoix

**Pull request merged**